o1vm/src/interpreters/mips/witness.rs

```diff
@@ -362,12 +362,26 @@ impl<Fp: Field, PreImageOracle: PreImageOracleT> InterpreterEnv for Env<Fp, PreI
     }
 
     fn equal(&mut self, x: &Self::Variable, y: &Self::Variable) -> Self::Variable {
-        // To avoid subtraction overflow in the witness interpreter for u32
-        if x > y {
-            self.is_zero(&(*x - *y))
-        } else {
-            self.is_zero(&(*y - *x))
-        }
+        // We replicate is_zero(x-y), but working on field elt,
+        // to avoid subtraction overflow in the witness interpreter for u32
+        let to_zero_test = Fp::from(*x) - Fp::from(*y);
+        let res = {
+            let pos = self.alloc_scratch();
+            let is_zero: u64 = if to_zero_test == Fp::zero() { 1 } else { 0 };
+            self.write_column(pos, is_zero);
+            is_zero
+        };
+        let _to_zero_test_inv_or_zero = {
+            let pos = self.alloc_scratch();
+            let inv_or_zero = if to_zero_test == Fp::zero() {
```

Comment:
You can probably call `inverse_or_zero` in these lines, no? I guess the bug comes from the `res` block, there's no need to unroll the `inv_or_zero` to solve the bug.

Reply:
I thought so at first as well, but Inverse_or_zero is forced to return a Variable by the signature, so we can't get the resulting field_inversion we desire here.